crates/wasm-encoder/src/reencode.rs

```diff
@@ -0,0 +1,1308 @@
+        data: &'a mut crate::DataSection,
+        datum: wasmparser::Data<'_>,
+    ) -> Result<&'a mut crate::DataSection> {
+        utils::parse_data(self, data, datum)
+    }
+
+    fn parse_element_section<'a>(
+        &mut self,
+        elements: &'a mut crate::ElementSection,
+        section: wasmparser::ElementSectionReader<'_>,
+    ) -> Result<&'a mut crate::ElementSection> {
+        utils::parse_element_section(self, elements, section)
+    }
+
+    fn parse_element<'a>(
+        &mut self,
+        elements: &'a mut crate::ElementSection,
+        element: wasmparser::Element<'_>,
+    ) -> Result<&'a mut crate::ElementSection> {
+        utils::parse_element(self, elements, element)
+    }
+
+    fn table_index(&mut self, table: u32) -> u32 {
+        utils::table_index(self, table)
+    }
+
+    fn global_index(&mut self, global: u32) -> u32 {
+        utils::global_index(self, global)
+    }
+
+    fn data_index(&mut self, data: u32) -> u32 {
+        utils::data_index(self, data)
+    }
+
+    fn element_index(&mut self, element: u32) -> u32 {
+        utils::element_index(self, element)
+    }
+
+    fn const_expr(&mut self, const_expr: wasmparser::ConstExpr) -> Result<crate::ConstExpr> {
+        utils::const_expr(self, const_expr)
+    }
+
+    fn block_type(&mut self, arg: wasmparser::BlockType) -> Result<crate::BlockType> {
+        utils::block_type(self, arg)
+    }
+
+    fn instruction<'a>(&mut self, arg: wasmparser::Operator<'a>) -> Result<crate::Instruction<'a>> {
+        utils::instruction(self, arg)
+    }
+
+    fn parse_code_section<'a>(
+        &mut self,
+        code: &'a mut crate::CodeSection,
+        section: wasmparser::CodeSectionReader<'_>,
+    ) -> Result<&'a mut crate::CodeSection> {
+        utils::parse_code_section(self, code, section)
+    }
+
+    fn parse_function_body<'a>(
+        &mut self,
+        code: &'a mut crate::CodeSection,
+        func: wasmparser::FunctionBody<'_>,
+    ) -> Result<&'a mut crate::CodeSection> {
+        utils::parse_function_body(self, code, func)
+    }
+
+    fn new_function_with_parsed_locals(
+        &mut self,
+        func: &wasmparser::FunctionBody<'_>,
+    ) -> Result<crate::Function> {
+        utils::new_function_with_parsed_locals(self, func)
+    }
+
+    fn parse_instruction<'a>(
+        &mut self,
+        function: &'a mut crate::Function,
+        reader: &mut wasmparser::OperatorsReader<'_>,
+    ) -> Result<&'a mut crate::Function> {
+        utils::parse_instruction(self, function, reader)
+    }
+}
+
+#[derive(Debug)]
+pub struct RoundtripReencoder;
+
+impl WasmParserToWasmEncoder for RoundtripReencoder {}
+
+pub mod utils {
+    use super::{Error, Result, WasmParserToWasmEncoder};
+
+    pub fn core_module<'a, 'b>(
+        reencoder: &mut (impl ?Sized + WasmParserToWasmEncoder),
+        module: &'a mut crate::Module,
+        sections: impl Iterator<Item = Result<wasmparser::Payload<'b>, wasmparser::BinaryReaderError>>,
```

Comment:
Instead of taking `sections` could this instead take `wasmparser::Parser` and `&[u8]`? There's various invariants around the state transitions of parsing the code section which I think want to be asserted on below which aren't guaranteed with an arbitrary iterator but are guaranteeed for a parser/slice combo

Reply:
Done